internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
-	for {
-		if bytes.Equal(tx, wtx.tx) {
-			// We've found a tx in the recheck list that matches the tx that we
-			// received from the ABCI application.
-			// Break, and use this transaction for further checks.
-			break
-		}
-
-		txmp.logger.Error(
-			"re-CheckTx transaction mismatch",
-			"got", wtx.tx.Hash(),
-			"expected", types.Tx(tx).Key(),
-		)
 
-		if txmp.recheckCursor == txmp.recheckEnd {
-			// we reached the end of the recheckTx list without finding a tx
-			// matching the one we received from the ABCI application.
-			// Return without processing any tx.
-			txmp.recheckCursor = nil
+	numLeft := atomic.AddInt64(&txmp.txRecheck, -1)
+	if numLeft <= 0 {
+		defer txmp.notifyTxsAvailable()
+		if numLeft < 0 {
 			return
 		}
-
-		txmp.recheckCursor = txmp.recheckCursor.Next()
-		wtx = txmp.recheckCursor.Value.(*WrappedTx)
 	}
 
-	// Only evaluate transactions that have not been removed. This can happen
-	// if an existing transaction is evicted during CheckTx and while this
-	// callback is being executed for the same evicted transaction.
-	if !txmp.txStore.IsTxRemoved(wtx.hash) {
-		var err error
-		if txmp.postCheck != nil {
-			err = txmp.postCheck(tx, checkTxRes.CheckTx)
-		}
-
-		if checkTxRes.CheckTx.Code == abci.CodeTypeOK && err == nil {
-			wtx.priority = checkTxRes.CheckTx.Priority
-		} else {
-			txmp.logger.Debug(
-				"existing transaction no longer valid; failed re-CheckTx callback",
-				"priority", wtx.priority,
-				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
-				"err", err,
-				"code", checkTxRes.CheckTx.Code,
-			)
-
-			if wtx.gossipEl != txmp.recheckCursor {
-				panic("corrupted reCheckTx cursor")
-			}
+	txmp.metrics.RecheckTimes.Add(1)
+	tx := types.Tx(req.GetCheckTx().Tx)
 
-			txmp.removeTx(wtx, !txmp.config.KeepInvalidTxsInCache)
-		}
+	// Find the transaction reported by the ABCI callback. It is possible the
+	// transaction was evicted during the recheck, in which case the transaction
+	// will be gone.
+	elt, ok := txmp.txByKey[tx.Key()]
+	if !ok {
+		return
 	}
+	wtx := elt.Value.(*WrappedTx)
 
-	// move reCheckTx cursor to next element
-	if txmp.recheckCursor == txmp.recheckEnd {
-		txmp.recheckCursor = nil
-	} else {
-		txmp.recheckCursor = txmp.recheckCursor.Next()
+	// If a postcheck hook is defined, call it before checking the result.
+	var err error
+	if txmp.postCheck != nil {
+		err = txmp.postCheck(tx, checkTxRes.CheckTx)
 	}
 
-	if txmp.recheckCursor == nil {
-		txmp.logger.Debug("finished rechecking transactions")
-
-		if txmp.Size() > 0 {
-			txmp.notifyTxsAvailable()
-		}
+	if checkTxRes.CheckTx.Code == abci.CodeTypeOK && err == nil {
+		wtx.SetPriority(checkTxRes.CheckTx.Priority)
+		return // N.B. Size of mempool did not change
 	}
 
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.Priority(),
+		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+		"err", err,
+		"code", checkTxRes.CheckTx.Code,
+	)
```

Comment:
Do we want to report other information about the response i.e. codespace or log?

Reply:
I don't know. If you have suggestions I'm glad to add them; I mainly just kept the logs as they were.